arrows/vxl/convert_image.cxx

```diff
@@ -0,0 +1,510 @@
+            static_cast< double >( std::numeric_limits< InputType >::max() );
+  }
+
+  auto const ni = src.ni();
+  auto const nj = src.nj();
+  auto const np = src.nplanes();
+  dst.set_size( ni, nj, np );
+
+  // Stretch image to upper and lower percentile bounds
+  vil_transform(
+    src, dst,
+    [ lower_bound, upper_bound, scale ]( InputType pixel ){
+      if( pixel < lower_bound )
+      {
+        return static_cast< OutputType >( 0 );
+      }
+      else if( pixel > upper_bound )
+      {
+        return static_cast< OutputType >(
+          std::numeric_limits< OutputType >::max() );
+      }
+      else
+      {
+        return static_cast< OutputType >(
+          static_cast< double >( pixel - lower_bound ) * scale );
+      }
+    } );
+}
+
+} // namespace <anonoymous>
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class convert_image::priv
+{
+public:
+  priv()
+    : format{ "byte" }
+      , single_channel{ false }
+      , scale_factor{ 0.0 }
+      , random_grayscale{ 0.0 }
+      , percentile_norm{ -1.0 }
+  {
+  }
+
+  // Convert a fraction of images to gray
+  template < typename Type >
+  vil_image_view< Type >
+  random_gray_conversion( vil_image_view< Type > const& src,
+                          double const random_fraction );
+
+  // Apply appropriate transforms
+  template < typename ipix_t >
+  vil_image_view< ipix_t >
+  apply_transforms( vil_image_view_base_sptr& view );
+
+  // Scale and convert the image
+  template < typename opix_t, typename ipix_t >
+  vil_image_view< opix_t >
+  scale_and_convert( vil_image_view< ipix_t >& input );
+
+  std::string format;
+  bool single_channel;
+  double scale_factor;
+  double random_grayscale;
+  double percentile_norm;
+
+  std::random_device random_device;
+  std::mt19937 engine{ random_device() };
+  std::uniform_real_distribution< double > dist{ 0.0, 1.0 };
+};
+
+// ----------------------------------------------------------------------------
+template < typename Type >
+vil_image_view< Type >
+convert_image::priv
+::random_gray_conversion( vil_image_view< Type > const& src,
+                          double const random_fraction )
+{
+  auto random_value = dist( engine );
+
+  if( random_value < random_fraction )
+  {
+    vil_image_view< Type > compressed;
+    combine_channels( src, compressed );
+
+    vil_image_view< Type > dst{ src.ni(), src.nj(), src.nplanes() };
+
+    for( unsigned p = 0; p < src.nplanes(); ++p )
+    {
+      vil_image_view< Type > output_plane = vil_plane( dst, p );
+      vil_copy_reformat( compressed, output_plane );
+    }
+    return dst;
+  }
+  else
+  {
+    return src;
+  }
+}
+
+// ----------------------------------------------------------------------------
+template < typename ipix_t >
+vil_image_view< ipix_t >
+convert_image::priv
+::apply_transforms( vil_image_view_base_sptr& view )
+{
+  vil_image_view< ipix_t > output;
```

Comment:
BTW, sorry for not noticing earlier, but should this maybe first check that the input isn't already one-channel?

If yes, it might also make sense to move `output` to only the `if( single_channel )` block and make everything else just `return`.